lib/multiplexer/wrappers.go

```diff
@@ -155,3 +157,49 @@ func (l *Listener) Close() error {
 	l.cancel()
 	return nil
 }
+
+// PROXYEnabledListener wraps provided listener and can receive and apply PROXY headers and then pass connection up the chain.
+type PROXYEnabledListener struct {
+	cfg Config
+	mux *Mux
+	net.Listener
+}
+
+// NewPROXYEnabledListener creates news instance of PROXYEnabledListener
+func NewPROXYEnabledListener(cfg Config) (*PROXYEnabledListener, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	mux, err := New(cfg) // Creating Mux to leverage protocol detection with PROXY headers
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	muxListener := mux.SSH()
+	go func() {
+		if err := mux.Serve(); err != nil && !utils.IsOKNetworkError(err) {
+			mux.Entry.WithError(err).Error("Mux encountered err serving")
+		}
+	}()
+	pl := &PROXYEnabledListener{
+		cfg:      cfg,
+		mux:      mux,
+		Listener: muxListener,
+	}
+
+	return pl, nil
+}
+
+func (p *PROXYEnabledListener) Close() error {
+	if err := p.mux.Close(); err != nil {
+		return err
+	}
+	return p.Listener.Close()
```

Comment:
If closing the artificial SSH listener is necessary then this ought to be a `trace.NewAggregate(p.mux.Close(), p.Listener.Close())` (but I think that just closing the mux itself should be fine?)

Reply:
Yep, with relying on mux's listener just closing mux itself is now enough, changed 👍